### PR TITLE
Visible CollisionShape3D faces and random Color

### PR DESCRIPTION
**Pull request by bitsycore**
edit : After lots of change https://github.com/godotengine/godot/pull/40123#issuecomment-655251208 better show the state of the PR

https://github.com/godotengine/godot-proposals/issues/906
It add a default color face to `BoxShape3D`, `CapsuleShape3D`, `CylinderShape3D` and `SphereShape3D` when used with `CollisionShape3D`. 
Each instance can have custom color and texture to differentiate them to help gamedesigning bigger scene.

rLMxdbfhPR

I still have thing TODO but not sure what fit the best for godot

- [x] Default Color in EditorSettings
- [x] Random Color
- ~~Miss other shapes~~
- ~~Random Color EditorSettings~~

### CODE REVIEW

**Comment by Coldragon** [comment]
I added random_color in the editor_settings, but when I try to get the value in the constructor to enable use_custom_color in the shape I have a mutex insult. It happen during the "docbase pass".

Any idea where I should check the editor value instead ?

**Comment by Coldragon** [comment]
The PR is ready on my side for the 4 shapes, it only miss one feature : 
- `EditorSettings` Random Color : If disabled on creation a shape will not use it's custom color (that is randomized on the constructor). But I can't use EDITOR_DEF on its constructor, I have mutex error during the doc generation phase. If anyone have an idea

**Comment by YeldhamDev** [comment]
Those should probably be wrapped inside `#ifdef TOOLS_ENABLED` sections, as there's no point having this code on release.

**Comment by Coldragon** [comment]
> Those should probably be wrapped inside `#ifdef TOOLS_ENABLED` sections, as there's no point having this code on release.

Oh true, for the Shape parts I guess ?

**Comment by YeldhamDev** [comment]
> Oh true, for the Shape parts I guess ? Gizmo are already tools ?

As far as I'm aware, yes. Anything within the "editor" folder is only compiled on editor builds already.

**Comment by Coldragon** [comment]
> > Oh true, for the Shape parts I guess ? Gizmo are already tools ?
> 
> As far as I'm aware, yes. Anything within the "editor" folder is only compiled on editor builds already.

Ok thank's, I didn't though about that. I'll do a little comit

**Comment by YeldhamDev** [comment]
Please amend the commits together: https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by Calinou** [comment]
How does this behave when using **Debug > Visible Collision shapes** during gameplay?

**Comment by Coldragon** [comment]
> How does this behave when using **Debug > Visible Collision shapes** during gameplay?

The behavior didn't changed for that part, I only added visibility inside CollisionShape3DGizmoPlugin

**Comment by Coldragon** [comment]
If I edit the get_debug_mesh and add the material creation directly inside shape it could be easly done

> > How does this behave when using **Debug > Visible Collision shapes** during gameplay?
> 
> The behavior didn't changed for that part, I only added visibility inside `CollisionShape3DGizmoPlugin`.
> 
> But it could be done if I add the mesh and material creation inside the shape directly and use it in get_debug_mesh()

**Comment by Zireael07** [comment]
In-editor and debug visible collision shapes should look the same imho.

**Comment by Coldragon** [comment]
Ok I added visible faces on runtime.
image
I will not squash the commit directly cause I want to have a bit of review on it cause i'm not sure if I did everything right

File: scene/resources/box_shape_3d.h
Comment (aaronfranke) [inline scene/resources/box_shape_3d.h:70]: Shouldn't these be defined on `Shape3D` rather than repeated for every shape?

**Review by aaronfranke — COMMENTED** [review]
What is the use case for custom textures? I personally wouldn't find that useful for debugging, but custom colors do seem nice. I'm not against custom textures, I'm just curious why this was done.

File: scene/resources/box_shape_3d.h
Comment (Coldragon) [reply scene/resources/box_shape_3d.h:70]: At first I did that but not all shape type are implemented

**Comment by Coldragon** [comment]
> What is the use case for custom textures? I personally wouldn't find that useful for debugging, but custom colors do seem nice. I'm not against custom textures, I'm just curious why this was done.

In source engine there were texture for trigger and you could have different one for "death", "spawn", "ladder" etc...
image

image

File: scene/resources/box_shape_3d.h
Comment (Coldragon) [reply scene/resources/box_shape_3d.h:70]: I could maybe add a CollisionShape3DVisual (abstract) and move everything on it ? But I'll still need to have the `ArrayMesh` generation in the child class cause' i need to know the kind of shape

Comment (aaronfranke) [reply scene/resources/box_shape_3d.h]: Which Shape3D types don't or can't have it implemented?

**Comment by Coldragon** [comment]
Ok I have a big problem with that PR. 
On release I remove the property added to the Shapes with DEBUG_ENABLED. 
So when I try to export a game with collision shapes it instant crash because I guess godot try to load property that doesn't exist in the .tscn ? Is there a way to do Editor Only property ?
Thank's

**Comment by Zylann** [comment]
There are a few things worrying me as well with this PR:
- The intermediate class now exists for the sole purpose of editor debugging. It's also meaningless for a game or exported.
- If the properties aren't properly stripped on export, a user setting up textures will end up with them being loaded in the final game, even if not used at all, which is a waste of resources.
Unless we consider it's fine, we may need better support for stripping editor-only features.

**Comment by Coldragon** [comment]
@Zylann Like you proposed in Discord, I could Give a random color on the shape in the gizmo plugin (to stay editor side, using it's ObjectID).
I would have used composition but I was not sure how to bind method using composition ?

**Comment by Coldragon** [comment]
I pushed a new version with random color based on `ObjectID` of the Shape and you can switch to a common color in the editor option.
image
image
image

- Changing the common color need to restart the editor.
- Disabling the "random color" need to trigger readraw() on the gizmo plugin, so selecting or other the shape.
- At runtime it only show random color when Visible Collision Shape is enable. Should I add a parameters at godot launch like Visible Collision Shape do ?

**Comment by RaTcHeT302** [comment]
OH MAN, MORE SOURCE ENGINE STYLE, HAMMER EDITOR CHANGES, I LOVE IT

hammer is such an easy to use tool, it's so stupid powerful, i'd love if the engine went towards the source engine's ease of use direction and easy of convenience

File: editor/node_3d_editor_gizmos.cpp
Comment (JFonS) [inline editor/node_3d_editor_gizmos.cpp:3642]: If I'm not mistaken, this code is repeated for every collision shape type. Can we have it just once at the top of the function and store the material in a variable that is then used on the various `add_mesh()` calls?

**Review by JFonS — COMMENTED** [review]
Looks mostly good, I left a review comment about some code duplication.

 I'm OK with not having custom textures, I don't think they added too much value. Per-instance custom colors could be useful, if anyone wants to work on that, but this PR should be good to merge regardless (after rebasing and removing duplicated code) :)

**Comment by reduz** [comment]
I feel the texture may be a bit much to be honest, but color could be ok.

File: scene/3d/collision_shape_3d.cpp
Comment (pouleyKetchoupp) [inline scene/3d/collision_shape_3d.cpp:232]: The logic for drawing shapes has been moved to `CollisionObject3D` and is now doing calls to the `RenderingServer` directly instead of adding mesh instances in the tree.

So after a rebase, `debug_shape_faces` will have to move to `CollisionObject3D`, see:
https://github.com/godotengine/godot/blob/ade4e9320a6ca403b8053fe5828d3f9ce809338c/scene/3d/collision_object_3d.cpp#L340

And also:
-Same as for the editor gizmo, it could use some simpler logic by making `get_debug_arraymesh_faces()` virtual just like `get_debug_mesh()` does.
-Options for faces enabled and custom colors can be check from project settings
`GLOBAL_GET("debug/shapes/collision/draw_3d_faces")`
`GLOBAL_GET("debug/shapes/collision/use_random_color_for_3d_faces")`
-It needs one more option in project settings to set the default color, defined in scene tree as well:
GLOBAL_DEF("debug/shapes/collision/shape_color_3d_faces", Color(0.0, 0.6, 0.7, 0.42));

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Sorry for the late review. Are you still interested in finalizing this PR? This is a welcome change, although it will need a bit of refactoring before it can be merged. See my comments for details.

As a note for reviewers:
The PR description is a bit misleading because of changes that happened over time, but there's no more texture option and no more settings in shape data.
Now it's just two general options for face colors and random face colors.

After these changes, it would be interesting to add an option for custom colors, but it would be probably in `CollisionShape` instead of the shape resource itself.

**Comment by Cold-0** [comment]
@pouleyKetchoupp Hello, yeah I would still be interested finalizing this PR. But I don't have time for now until end of september (or at least can't promise anything).

I took note of the change you requested and thanks for these comment (also JFonS), if that PR is still needed after September, I will finalize it.

For the custom color per instance I'll try adding it too.

**Comment by pouleyKetchoupp** [comment]
@Cold-0 Sounds good! No problem to wait until end of September.

File: editor/node_3d_editor_gizmos.cpp
Comment (Calinou) [reply editor/node_3d_editor_gizmos.cpp:3631]: Should random colors even be kept in the first place? We don't use random colors for raycasts or 2D collision shapes after all.

**Comment by Cold-0** [comment]
When the "custom color" would be (re)added, when the collision node is
added, we add a random color and if the "use custom color" is check use it.
And maybe a config in the editor to allow the "use custom color" checked on
node creation ?

Le sam. 4 sept. 2021 à 00:51, Hugo Locurcio ***@***.***> a
écrit :

File: editor/node_3d_editor_gizmos.cpp
Comment (pouleyKetchoupp) [reply editor/node_3d_editor_gizmos.cpp:3631]: It could be simplified if it was just a default color and custom colors, but depending how random colors are implemented, it could be pretty simple (and it could be for both 2D and 3D eventually).

Quoting the last message from @Cold-0:
> When the "custom color" would be (re)added, when the collision node is
added, we add a random color and if the "use custom color" is check use it.
And maybe a config in the editor to allow the "use custom color" checked on
node creation ?

We could even make custom color as simple as for Raycast3D and make the value black by default, which means it uses the editor/runtime default.
Then we could have editor settings to generate random custom colors on 2D/3D collision shape creation.

@Calinou Do you think one of these solutions could be ok, since only editor code would be affected? Some users might find it practical for debugging.

**Comment by YuriSizov** [comment]
Seems this PR was automatically closed because the author deleted their branch/entire fork. This is probably salvageable and given there has been plenty of activity on this PR, there is interest.

File: scene/3d/collision_shape_3d.cpp
Comment (Calinou) [reply scene/3d/collision_shape_3d.cpp:232]: Also, I noticed a material is created for every instance. This will be a bottleneck in large scenes.

Instead, vertex colors should be used to tint each shape by giving it vertex colors when generating the mesh, and displaying those as albedo (similar to what is done for Light3D gizmo lines and billboard icons).

**Comment by Zireael07** [comment]
This needs to be salvaged.

**Comment by Calinou** [comment]
> What is the use case for custom textures? I personally wouldn't find that useful for debugging, but custom colors do seem nice. I'm not against custom textures, I'm just curious why this was done.

Regarding the use case for custom textures (or even a single texture for all shapes), one good reason to have them is that they give you spatial awareness, which a plain color doesn't. In short, seeing a textured pattern (even a checkerboard or grid) will give you a visual indicator of how far away the shape is from the camera (if it's a huge shape covering a large plane, for instance). This also helps to distinguish intersecting shapes of the same color if the outline isn't visible for some reason.

I'm putting this here in case someone is interested in salvaging this PR, although this can be implemented in a separate future PR.

**Comment by bitsydoge** [comment]
image
This can also be done procedurally I think.
Color, Grid, Text, Can easly give these nice visual

**Pull request closed without merging**
